File: ios/Plugin/UsercentricsCmpPlugin.swift
import Foundation
import Capacitor
import Usercentrics
import UsercentricsUI

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(UsercentricsCmpPlugin)
public class UsercentricsCmpPlugin: CAPPlugin {
    private var _call: CAPPluginCall?
    private var usercentricsInitialized: Bool?

    @objc(init:)
    func initialize(_ call: CAPPluginCall) {
        self._call = call

        if ((self.usercentricsInitialized) == nil) {
            self.configureUsercentrics()
        }

        DispatchQueue.main.async {
            UsercentricsCore.isReady { [weak self] status in
                guard let self = self else { return }
                if status.shouldCollectConsent {
                    self.presentFirstLayer()
                } else {
                    // Apply stored consent (Don't show CMP)
                    self.applyConsent(with: status.consents)
                }
            } onFailure: { error in
                self._call?.reject(error.localizedDescription)
            }
        }
    }

    @objc func update(_ call: CAPPluginCall) {
        self._call = call

        if ((self.usercentricsInitialized) == nil) {
            self._call?.reject("Usercentrics not configured yet. Ensure init function has beeen called before")

            return;
        }
        
        UsercentricsCore.shared.changeLanguage(language: self._call?.getString("language") ?? "en") {
            DispatchQueue.main.async {
                self.presentSecondLayer()
            }
        } onFailure: { error in
            DispatchQueue.main.async {
                // ignore language switch error, present cmp anyway
                self.presentSecondLayer()
            }
        }
    }

    @objc func reset(_ call: CAPPluginCall) {
        self._call = call

        if ((self.usercentricsInitialized) == nil) {
            self._call?.reject("Usercentrics not configured yet. Ensure init function has beeen called before")

            return;
        }

        DispatchQueue.main.async {
            UsercentricsCore.reset();

            self.usercentricsInitialized = nil;
            self._call?.resolve(["vendors": []])
        }
    }

    private func configureUsercentrics() {
        guard let settingsId = self._call?.getString("settingsId") else {
            self._call?.reject("settingsId missing")
            return
        }

        let options = UsercentricsOptions(settingsId: settingsId)
        // options.loggerLevel = .debug
        options.defaultLanguage = self._call?.getString("language") ?? "en"
        UsercentricsCore.configure(options: options)
        self.usercentricsInitialized = true;
    }
    
    
    private func presentFirstLayer() {
        guard let navigationController = self.bridge?.viewController else { fatalError("Navigation Controller needed") }
        
        let bannerSettings = BannerSettings(firstLayerStyleSettings: FirstLayerStyleSettings(layout: .popup(position: .center)))

        // Launch Usercentrics Banner with your settings
        let banner = UsercentricsBanner(bannerSettings: bannerSettings)
        banner.showFirstLayer(hostView: navigationController) { [weak self] response in
            guard let self = self else { return }
            /// Process consents
            self.applyConsent(with: response.consents)
        }
    }
    
    private func presentSecondLayer() {
        guard let navigationController = self.bridge?.viewController else { fatalError("Navigation Controller needed") }

        // This is useful when you need to call our CMP from settings screen for instance, therefore the user may dismiss the view
        let banner = UsercentricsBanner(bannerSettings: BannerSettings(secondLayerStyleSettings: SecondLayerStyleSettings(showCloseButton: true)))
        banner.showSecondLayer(hostView: navigationController) { [weak self] response in
            guard let self = self else { return }
            /// Process consents
            self.applyConsent(with: response.consents)
        }
    }

    private func applyConsent(with consents: [UsercentricsServiceConsent]) {
        var vendors: [[String: Any]] = []

        for vendor in consents {
            let mappedVendor: [String: Any] = [
                "id": vendor.templateId,
                "type": vendor.type?.name as Any,
                "status": vendor.status,
                "label": vendor.dataProcessor,
                "version": vendor.version,
            ]

            vendors.append(mappedVendor)
        }

        print("applied consents", consents);

        self._call?.resolve(["vendors": vendors])
    }
}

File: android/src/main/java/com/getcapacitor/community/usercentrics/UsercentricsCmpPlugin.kt
package com.getcapacitor.community.usercentrics

import android.util.Log
import androidx.activity.result.ActivityResultLauncher
import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import com.usercentrics.sdk.*

@CapacitorPlugin(
  name = "UsercentricsCmp",
  permissions = []
)
class UsercentricsCmpPlugin : Plugin() {
  private val TAG = "UsercentricsCmp"

  private var banner: UsercentricsBanner? = null
  private var callbackId: String? = null
  private var settingsId: String? = null
  private var usercentricsInitialized: Boolean = false

  override fun load() {
    super.load()
  }

  @PluginMethod()
  fun init(call: PluginCall) {
    settingsId = call.getString("settingsId", "")

    if (settingsId == "") {
      call.reject("settingsId missing")
      return
    }
    callbackId = call.callbackId
    bridge.saveCall(call)

    val options = UsercentricsOptions(settingsId = settingsId!!)
    options.defaultLanguage = call.getString("language", "en")!!

    if (!usercentricsInitialized) {
      Usercentrics.initialize(context = context, options = options)
      usercentricsInitialized = true
    }

    Usercentrics.isReady(onSuccess = { status ->
      if (status.shouldCollectConsent) {
        showFirstLayer()
      } else {
        applyConsent(status.consents)
      }
    }, onFailure = {
      // Handle error
      Log.e(TAG, it.toString())
      call.reject("error")
      call.release(bridge)
    })
  }

  @PluginMethod
  fun update(call: PluginCall) {
    if (!usercentricsInitialized) {
      call.reject("Usercentrics not initialized yet. Ensure init function has beeen called before")

      return
    }

    callbackId = call.callbackId
    bridge.saveCall(call)

    Log.d(TAG, "update cmp")

    Usercentrics.instance.changeLanguage(call.getString("language", "en")!!, onSuccess = {
      showSecondLayer()
    }, onFailure = {
      // ignore language switch error, present cmp anyway
      Log.e(TAG, it.toString())
      showSecondLayer()
    })
  }

  @PluginMethod
  fun reset(call: PluginCall) {
    if (!usercentricsInitialized) {
      call.reject("Usercentrics not initialized yet. Ensure init function has beeen called before")

      return
    }

    callbackId = call.callbackId
    bridge.saveCall(call)
    Log.d(TAG, "reset cmp")
    Usercentrics.reset()
    usercentricsInitialized = false

    val consents = JSObject()
    consents.put("vendors", JSArray())

    call.resolve(consents)
    call.release(bridge)
  }

  private fun showFirstLayer(settings: BannerSettings? = null) {

    // Launch Usercentrics Banner with your settings
    banner = UsercentricsBanner(context, settings).also {
      it.showFirstLayer(
        callback = ::handleUserResponse
      )
    }
  }

  private fun showSecondLayer() {
    // This is useful when you need to call our CMP from settings screen for instance, therefore the user may dismiss the view
    val settings = BannerSettings(
      secondLayerStyleSettings = SecondLayerStyleSettings(
        showCloseButton = true,
      )
    )


    activity.runOnUiThread {
      banner = UsercentricsBanner(context, settings).also {
        it.showSecondLayer(
          callback = ::handleUserResponse
        )
      }
    }
  }

  private fun handleUserResponse(userResponse: UsercentricsConsentUserResponse?) {
    userResponse ?: return

    applyConsent(userResponse.consents)
  }

  private fun applyConsent(consents: List<UsercentricsServiceConsent>?) {
    val call = bridge.getSavedCall(callbackId)
    val vendors = arrayListOf<JSObject>()

    consents?.forEach { vendor ->
      val jsVendor = JSObject()

      jsVendor.put("id", vendor.templateId)
      jsVendor.put("type", vendor.type.toString())
      jsVendor.put("status", vendor.status)
      jsVendor.put("label", vendor.dataProcessor)
      jsVendor.put("version", vendor.version)

      vendors.add(jsVendor)
      Log.d(TAG, "vendor: ${vendor.templateId} (${vendor.dataProcessor})")
    }

    val consents = JSObject()
    consents.put("vendors", JSArray(vendors))

    call.resolve(consents)
    call.release(bridge)
  }
}
